### PR TITLE
Add LifecycleRunnable

### PR DESCRIPTION
**Pull request by pickypg**
I've noticed throughout some of the code that we have a need to remove the boilerplate lifecycle check when starting/rescheduling certain runnables. This provides a simpler implementation to get this functionality without duplicating it.

It also helps to avoid constantly rescheduling in some cases.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * @param lifecycle The lifecycle to react too
+     * @param logger The logger to use when logging
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public AbstractLifecycleRunnable(Lifecycle lifecycle, ESLogger logger) {
+        Objects.requireNonNull(lifecycle, "lifecycle must not be null");
+        Objects.requireNonNull(logger, "logger must not be null");
+
+        this.lifecycle = lifecycle;
```
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java]: `Objects.requireNonNull()` either returns the first argument or throws an exception so you could simplify this to:

```java
this.lifecycle = Objects.requireNonNull(lifecycle, "lifecycle must not be null");
this.logger = Objects.requireNonNull(logger, "logger must not be null");
```

File: core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+     * {@inheritDoc}
+     * <p>
+     * This invokes {@link #doRunInLifecycle()} <em>only</em> if the {@link #lifecycle} is not stopped or closed. Otherwise it exits
+     * immediately.
+     */
+    @Override
+    protected final void doRun() throws Exception {
+        // prevent execution if the service is stopped
+        if (lifecycle.stoppedOrClosed()) {
+            logger.trace("service is stopping. exiting");
```
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java]: I am not sure whether the log message is too specific, i.e. the subclass must not necessarily be a service.

**Comment by danielmitterdorfer** [comment]
Left a few minor comments, otherwise LGTM.

File: core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java (hunk 1 above)
Comment (pickypg) [inline core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java]: I prefer that style, but I searched around the code and saw the above style so I went to match it. Looks like I didn't look deep enough. I'll switch it.

File: core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java (hunk 2 above)
Comment (pickypg) [inline core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java]: Have any preferred name? Perhaps just changed it to lifecycle is stopping? I think that service is always going to be associated with a lifecycle, but given that it's trace logging I'll just switch it to:

> lifecycle is stopping. exiting

**Comment by pickypg** [comment]
Closed by f2aaa28362b22ba325bc846e633ee5e41b32135f

**Pull request closed without merging**